prov/gni/test/smrn.c

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright (c) 2017 Cray Inc. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+
+#include <stdio.h>
+#include <stdbool.h>
+#include <sys/mman.h>
+#include "gnix_smrn.h"
+
+#include <criterion/criterion.h>
+#include <criterion/logging.h>
+#include "common.h"
+
+#define GNIX_DEFAULT_RQ_CNT 4
+
+static struct gnix_smrn *smrn;
+static struct gnix_smrn_rq *rqs[GNIX_DEFAULT_RQ_CNT];
+static void **memory_regions;
+
+static void smrn_setup(void)
+{
+	int ret;
+	int i;
+	struct fi_info *info;
+
+	ret = fi_getinfo(FI_VERSION(1, 5), NULL, NULL, 0, NULL, &info);
+	cr_assert_eq(ret, FI_SUCCESS);
+
+	fi_freeinfo(info);
+
+	ret = _gnix_smrn_init();
+	cr_assert_eq(ret, 0, "_gnix_smrn_init failed, ret=%d\n", ret);
+
+	ret = _gnix_smrn_open(&smrn);
+	cr_assert(ret == 0, "_gnix_smrn_open failed");
+
+	for (i = 0; i < GNIX_DEFAULT_RQ_CNT; i++) {
+		rqs[i] = calloc(1, sizeof(*rqs[i]));
+		cr_assert_neq(rqs[i], NULL);
+
+		fastlock_init(&rqs[i]->lock);
+		dlist_init(&rqs[i]->list);
+		dlist_init(&rqs[i]->entry);
+	}
+}
+
+static void smrn_teardown(void)
+{
+	int ret;
+	int i;
+
+	ret = _gnix_smrn_close(smrn);
+	cr_assert(ret == 0, "_gnix_smrn_close failed");
+
+	for (i = 0; i < GNIX_DEFAULT_RQ_CNT; i++) {
+		free(rqs[i]);
+		rqs[i] = NULL;
+	}
+}
+
+TestSuite(smrn,
+	  .init = smrn_setup,
+	  .fini = smrn_teardown);
+
+#define RQ_ENTRIES 21
+#define REGIONS (GNIX_DEFAULT_RQ_CNT * RQ_ENTRIES)
+struct test_structure {
+	struct gnix_smrn_context context;
+	int pending;
+};
+
+Test(smrn, simple)
```

Comment:
Actually, it would be more work than it would be worth to try to combine them. 

The simple test strides the memory access in a manner that should be similar to the threaded test, but I cannot combine the functions to get the same behavior because of the striding.